controllers/twophase/state_machine.go

```diff
@@ -0,0 +1,296 @@
+	scheduler := m.Chaos.GetScheduler()
+	if scheduler == nil {
+		return nil, nil, errors.Errorf("misdefined scheduler")
+	}
+	m.Log.Info("iterate nextStart and nextRecover", "startTime", startTime, "duration", duration, "scheduler", scheduler)
+	nextStart, err := utils.NextTime(*scheduler, startTime)
+
+	if err != nil {
+		m.Log.Error(err, "failed to get the next start time")
+		return nil, nil, err
+	}
+	nextRecover := startTime.Add(duration)
+
+	counter := 0
+	// if the duration is too long, `nextRecover` could be after `nextStart`
+	// we can jump over a start to make sure `nextRecover` is before `nextStart`
+	for nextRecover.After(*nextStart) {
+		nextStart, err = utils.NextTime(*scheduler, *nextStart)
+		if err != nil {
+			m.Log.Error(err, "failed to get the next start time")
+			return nil, nil, err
+		}
+
+		counter++
+		if counter > iterMax {
+			err = errors.Errorf("the number of iterations exceeded with nextRecover(%s) nextStart(%s)", nextRecover, nextStart)
+			return nil, nil, err
+		}
+	}
+
+	return nextStart, &nextRecover, nil
+}
+
+var phaseTransitionMap = map[v1alpha1.ExperimentPhase]map[v1alpha1.ExperimentPhase]func(ctx context.Context, m *chaosStateMachine, targetPhase v1alpha1.ExperimentPhase, now time.Time) (bool, error){
+	v1alpha1.ExperimentPhaseUninitialized: {
+		v1alpha1.ExperimentPhaseUninitialized: none,
+		v1alpha1.ExperimentPhaseRunning:       apply,
+		v1alpha1.ExperimentPhaseWaiting:       none,
+		v1alpha1.ExperimentPhasePaused:        none,
+		v1alpha1.ExperimentPhaseFailed:        unexpected,
+		v1alpha1.ExperimentPhaseFinished:      none,
+	},
+	v1alpha1.ExperimentPhaseRunning: {
+		v1alpha1.ExperimentPhaseUninitialized: unexpected,
+		v1alpha1.ExperimentPhaseRunning:       none,
+		v1alpha1.ExperimentPhaseWaiting:       recover,
+		v1alpha1.ExperimentPhasePaused:        recover,
+		v1alpha1.ExperimentPhaseFailed:        unexpected,
+		v1alpha1.ExperimentPhaseFinished:      recover,
+	},
+	v1alpha1.ExperimentPhaseWaiting: {
+		v1alpha1.ExperimentPhaseUninitialized: unexpected,
+		v1alpha1.ExperimentPhaseRunning:       apply,
+		v1alpha1.ExperimentPhaseWaiting:       none,
+		v1alpha1.ExperimentPhasePaused:        none,
+		v1alpha1.ExperimentPhaseFailed:        unexpected,
+		v1alpha1.ExperimentPhaseFinished:      none,
+	},
+	v1alpha1.ExperimentPhasePaused: {
+		v1alpha1.ExperimentPhaseUninitialized: unexpected,
+		v1alpha1.ExperimentPhaseRunning:       resume,
+		v1alpha1.ExperimentPhaseWaiting:       resume,
+		v1alpha1.ExperimentPhasePaused:        none,
+		v1alpha1.ExperimentPhaseFailed:        unexpected,
+		v1alpha1.ExperimentPhaseFinished:      none,
+	},
+	v1alpha1.ExperimentPhaseFailed: {
+		v1alpha1.ExperimentPhaseUninitialized: unexpected,
+		v1alpha1.ExperimentPhaseRunning:       apply,
+		v1alpha1.ExperimentPhaseWaiting:       none,
+		v1alpha1.ExperimentPhasePaused:        none,
+		v1alpha1.ExperimentPhaseFailed:        none,
+		v1alpha1.ExperimentPhaseFinished:      none,
```

Comment:
Actually `ExperimentPhaseFailed` means failing to apply. Failing to recover will keep the experiment in `ExperimentPhaseRunning`.

Maybe I should modify the name. I don't know whether this modification would break the dashboard, so I will try to modify it in another PR.

Reply:
OK 😁